### ISSUE TITLE
Truncate table does not work if static column is present

### ISSUE THREAD

**Issue by masumsoft** [strapdata/elassandra#132]
**Elasticsearch version**: 5.5.0

**Plugins installed**: []

**JVM version**: 1.8.0_101

**OS version**: macos Sierra

**Description of the problem including expected versus actual behavior**:

I'm having trouble with a bizarre issue. 

Say there is a static column in a table named `table_static`. Now another table named `table_normal` under the same keyspace has defined a dynamic mapping with discover. Now say `table_normal` required to alter a field to add/remove it. To do so, the elassandra index was deleted, then `table_normal` was altered, then index and mapping for `table_normal` was recreated. From now on truncate table starts to fail mysteriously for `table_static` that has a static column. Even if the `table_static` rows is deleted using a delete query and then truncate query again makes the deleted rows reappear!

**Steps to reproduce**:
 1. create a table with a static column
 2. insert some rows in it
 3. create another normal table
 4. define an elassandra index and a dynamic mapping using discover for the normal table.
 5. delete the index and alter the normal table to add a new column
 6. recreate the index and mapping using discover for the normal table
 7. try truncate table query for the static column table.
 8. does not work
 9. try deleting some rows from the static column table
 10. try truncate again. the deleted rows will reappear!

**Comment by masumsoft** [comment]
When truncate does not work, I get the following error log in elassandra running with logging enabled, though not sure if this is related to the issue:

```
2017-10-30 18:29:07,516 INFO  [MutationStage-1] org.apache.cassandra.db.ColumnFamilyStore.truncateBlocking(ColumnFamilyStore.java:2179) Truncating express_cassandra_tests_kspc1.event
2017-10-30 18:29:07,521 ERROR [PerDiskMemtableFlushWriter_0:210] org.apache.cassandra.service.CassandraDaemon$2.uncaughtException(CassandraDaemon.java:231) Exception in thread Thread[PerDiskMemtableFlushWriter_0:210,5,main]
java.lang.NoClassDefFoundError: com/carrotsearch/hppc/LongOpenHashSet
	at org.apache.cassandra.index.sasi.disk.DynamicTokenTreeBuilder.add(DynamicTokenTreeBuilder.java:52)
	at org.apache.cassandra.index.sasi.disk.OnDiskIndexBuilder.add(OnDiskIndexBuilder.java:186)
	at org.apache.cassandra.index.sasi.disk.PerSSTableIndexWriter$Index.add(PerSSTableIndexWriter.java:234)
	at org.apache.cassandra.index.sasi.disk.PerSSTableIndexWriter.lambda$nextUnfilteredCluster$0(PerSSTableIndexWriter.java:128)
	at java.util.HashMap.forEach(HashMap.java:1288)
	at org.apache.cassandra.index.sasi.disk.PerSSTableIndexWriter.nextUnfilteredCluster(PerSSTableIndexWriter.java:120)
	at org.apache.cassandra.db.ColumnIndex.lambda$add$1(ColumnIndex.java:250)
	at java.lang.Iterable.forEach(Iterable.java:75)
	at org.apache.cassandra.db.ColumnIndex.add(ColumnIndex.java:250)
	at org.apache.cassandra.db.ColumnIndex.buildRowIndex(ColumnIndex.java:111)
	at org.apache.cassandra.io.sstable.format.big.BigTableWriter.append(BigTableWriter.java:173)
	at org.apache.cassandra.io.sstable.SimpleSSTableMultiWriter.append(SimpleSSTableMultiWriter.java:48)
	at org.apache.cassandra.db.Memtable$FlushRunnable.writeSortedContents(Memtable.java:484)
	at org.apache.cassandra.db.Memtable$FlushRunnable.call(Memtable.java:519)
	at org.apache.cassandra.db.Memtable$FlushRunnable.call(Memtable.java:406)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at org.apache.cassandra.concurrent.NamedThreadFactory.lambda$threadLocalDeallocator$0(NamedThreadFactory.java:81)
	at java.lang.Thread.run(Thread.java:745)
2017-10-30 18:29:07,565 ERROR [MutationStage-1] org.apache.cassandra.db.TruncateVerbHandler.doVerb(TruncateVerbHandler.java:44) Error in truncation
java.lang.RuntimeException: java.util.concurrent.ExecutionException: java.lang.RuntimeException: java.util.concurrent.ExecutionException: java.lang.NoClassDefFoundError: com/carrotsearch/hppc/LongOpenHashSet
	at org.apache.cassandra.utils.FBUtilities.waitOnFuture(FBUtilities.java:404)
	at org.apache.cassandra.db.ColumnFamilyStore.forceBlockingFlush(ColumnFamilyStore.java:986)
	at org.apache.cassandra.db.ColumnFamilyStore.truncateBlocking(ColumnFamilyStore.java:2186)
	at org.apache.cassandra.db.TruncateVerbHandler.doVerb(TruncateVerbHandler.java:40)
	at org.apache.cassandra.net.MessageDeliveryTask.run(MessageDeliveryTask.java:66)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at org.apache.cassandra.concurrent.AbstractLocalAwareExecutorService$FutureTask.run(AbstractLocalAwareExecutorService.java:162)
	at org.apache.cassandra.concurrent.AbstractLocalAwareExecutorService$LocalSessionFutureTask.run(AbstractLocalAwareExecutorService.java:134)
	at org.apache.cassandra.concurrent.SEPWorker.run(SEPWorker.java:109)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.util.concurrent.ExecutionException: java.lang.RuntimeException: java.util.concurrent.ExecutionException: java.lang.NoClassDefFoundError: com/carrotsearch/hppc/LongOpenHashSet
	at java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.util.concurrent.FutureTask.get(FutureTask.java:192)
	at org.apache.cassandra.utils.FBUtilities.waitOnFuture(FBUtilities.java:400)
	... 9 common frames omitted
Caused by: java.lang.RuntimeException: java.util.concurrent.ExecutionException: java.lang.NoClassDefFoundError: com/carrotsearch/hppc/LongOpenHashSet
	at org.apache.cassandra.utils.Throwables.maybeFail(Throwables.java:51)
	at org.apache.cassandra.utils.FBUtilities.waitOnFutures(FBUtilities.java:392)
	at org.apache.cassandra.db.ColumnFamilyStore$Flush.flushMemtable(ColumnFamilyStore.java:1158)
	at org.apache.cassandra.db.ColumnFamilyStore$Flush.run(ColumnFamilyStore.java:1109)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at org.apache.cassandra.concurrent.NamedThreadFactory.lambda$threadLocalDeallocator$0(NamedThreadFactory.java:81)
	... 1 common frames omitted
Caused by: java.util.concurrent.ExecutionException: java.lang.NoClassDefFoundError: com/carrotsearch/hppc/LongOpenHashSet
	at java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.util.concurrent.FutureTask.get(FutureTask.java:192)
	at org.apache.cassandra.utils.FBUtilities.waitOnFutures(FBUtilities.java:385)
	... 6 common frames omitted
Caused by: java.lang.NoClassDefFoundError: com/carrotsearch/hppc/LongOpenHashSet
	at org.apache.cassandra.index.sasi.disk.DynamicTokenTreeBuilder.add(DynamicTokenTreeBuilder.java:52)
	at org.apache.cassandra.index.sasi.disk.OnDiskIndexBuilder.add(OnDiskIndexBuilder.java:186)
	at org.apache.cassandra.index.sasi.disk.PerSSTableIndexWriter$Index.add(PerSSTableIndexWriter.java:234)
	at org.apache.cassandra.index.sasi.disk.PerSSTableIndexWriter.lambda$nextUnfilteredCluster$0(PerSSTableIndexWriter.java:128)
	at java.util.HashMap.forEach(HashMap.java:1288)
	at org.apache.cassandra.index.sasi.disk.PerSSTableIndexWriter.nextUnfilteredCluster(PerSSTableIndexWriter.java:120)
	at org.apache.cassandra.db.ColumnIndex.lambda$add$1(ColumnIndex.java:250)
	at java.lang.Iterable.forEach(Iterable.java:75)
	at org.apache.cassandra.db.ColumnIndex.add(ColumnIndex.java:250)
	at org.apache.cassandra.db.ColumnIndex.buildRowIndex(ColumnIndex.java:111)
	at org.apache.cassandra.io.sstable.format.big.BigTableWriter.append(BigTableWriter.java:173)
	at org.apache.cassandra.io.sstable.SimpleSSTableMultiWriter.append(SimpleSSTableMultiWriter.java:48)
	at org.apache.cassandra.db.Memtable$FlushRunnable.writeSortedContents(Memtable.java:484)
	at org.apache.cassandra.db.Memtable$FlushRunnable.call(Memtable.java:519)
	at org.apache.cassandra.db.Memtable$FlushRunnable.call(Memtable.java:406)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	... 4 common frames omitted
2017-10-30 18:29:07,565 ERROR [MemtablePostFlush:109] org.apache.cassandra.service.CassandraDaemon$2.uncaughtException(CassandraDaemon.java:231) Exception in thread Thread[MemtablePostFlush:109,5,main]
java.lang.RuntimeException: java.util.concurrent.ExecutionException: java.lang.NoClassDefFoundError: com/carrotsearch/hppc/LongOpenHashSet
	at org.apache.cassandra.utils.Throwables.maybeFail(Throwables.java:51)
	at org.apache.cassandra.utils.FBUtilities.waitOnFutures(FBUtilities.java:392)
	at org.apache.cassandra.db.ColumnFamilyStore$Flush.flushMemtable(ColumnFamilyStore.java:1158)
	at org.apache.cassandra.db.ColumnFamilyStore$Flush.run(ColumnFamilyStore.java:1109)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at org.apache.cassandra.concurrent.NamedThreadFactory.lambda$threadLocalDeallocator$0(NamedThreadFactory.java:81)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.util.concurrent.ExecutionException: java.lang.NoClassDefFoundError: com/carrotsearch/hppc/LongOpenHashSet
	at java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.util.concurrent.FutureTask.get(FutureTask.java:192)
	at org.apache.cassandra.utils.FBUtilities.waitOnFutures(FBUtilities.java:385)
	... 6 common frames omitted
Caused by: java.lang.NoClassDefFoundError: com/carrotsearch/hppc/LongOpenHashSet
	at org.apache.cassandra.index.sasi.disk.DynamicTokenTreeBuilder.add(DynamicTokenTreeBuilder.java:52)
	at org.apache.cassandra.index.sasi.disk.OnDiskIndexBuilder.add(OnDiskIndexBuilder.java:186)
	at org.apache.cassandra.index.sasi.disk.PerSSTableIndexWriter$Index.add(PerSSTableIndexWriter.java:234)
	at org.apache.cassandra.index.sasi.disk.PerSSTableIndexWriter.lambda$nextUnfilteredCluster$0(PerSSTableIndexWriter.java:128)
	at java.util.HashMap.forEach(HashMap.java:1288)
	at org.apache.cassandra.index.sasi.disk.PerSSTableIndexWriter.nextUnfilteredCluster(PerSSTableIndexWriter.java:120)
	at org.apache.cassandra.db.ColumnIndex.lambda$add$1(ColumnIndex.java:250)
	at java.lang.Iterable.forEach(Iterable.java:75)
	at org.apache.cassandra.db.ColumnIndex.add(ColumnIndex.java:250)
	at org.apache.cassandra.db.ColumnIndex.buildRowIndex(ColumnIndex.java:111)
	at org.apache.cassandra.io.sstable.format.big.BigTableWriter.append(BigTableWriter.java:173)
	at org.apache.cassandra.io.sstable.SimpleSSTableMultiWriter.append(SimpleSSTableMultiWriter.java:48)
	at org.apache.cassandra.db.Memtable$FlushRunnable.writeSortedContents(Memtable.java:484)
	at org.apache.cassandra.db.Memtable$FlushRunnable.call(Memtable.java:519)
	at org.apache.cassandra.db.Memtable$FlushRunnable.call(Memtable.java:406)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	... 4 common frames omitted
```

**Comment by vroyer** [comment]
Hi,

You’ve got java.lang.NoClassDefFoundError for com/carrotsearch/hppc/LongOpenHashSet on a SASI index.
Elasticsearch requires com.carrotsearch:hppc:0.7.1 while cassandra SASI relies on hppc-0.5.4, and LongOpenHashSet has been removed from com.carrotsearch:hppc:0.7.1 !
I raised a ticket to the HPPC project about this issue http://issues.carrot2.org/browse/HPPC-168 <http://issues.carrot2.org/browse/HPPC-168>
For now, SASI index are not compatible with Elassandra.

Thanks fro reporting the issue.

**Comment by masumsoft** [comment]
Great, I've just removed the sasi indexes and the problem is now gone! Do you think if the hppc issue is resolved, then we can expect elassandra to work in harmony with the cassandra sasi indexes? AFAIK SASI is much performant for like queries than a lucene/elastic index. So it would be great if elassandra become compatible with SASI.

**Comment by masumsoft** [comment]
Also as the problem seems to be related to a class rename, is it feasible for elassandra to say modify the class names used by sasi in cassandra source code to use hppc 0.7.1?

**Comment by vroyer** [comment]
Yes, it’s just a naming issue, and we will upgrade hppc to 0.7.1 for cassandra, and release soon an update for elassandra.

**Comment by vroyer** [comment]
fixed by version 5.5.0.5.
